src/libsystemd/sd-device/sd-device.c

```diff
@@ -1473,34 +1473,58 @@ int device_add_tag(sd_device *device, const char *tag, bool both) {
         return 0;
 }
 
+static char *prefix_dev(const char *p) {
+        assert(p);
+
+        if (path_startswith(p, "/dev/"))
+                return strdup(p);
+
+        return path_join("/dev/", p);
+}
+
 int device_add_devlink(sd_device *device, const char *devlink) {
+        char *p;
         int r;
 
         assert(device);
         assert(devlink);
 
-        r = set_put_strdup(&device->devlinks, devlink);
+        if (!path_is_safe(devlink))
+                return -EINVAL;
+
+        p = prefix_dev(devlink);
+        if (!p)
+                return -ENOMEM;
+
+        path_simplify(p);
```

Comment:
`path_is_safe()` rejects `..`, but otherwise accepts everything except empty strings and paths exceeding length limits. And `prefix_dev()` accepts any string, incl. absolute paths. This seems iffy: we shouldn't take an absolute path and silently make it a path under `/dev/`. We should reject those instead.

Also, adding the `/dev/` prefix everywhere seems wasteful: all paths in the hashmap will have `/dev/`. And in fact all rules on my machine use relative paths in SYMLINK. So we'd be reallocating and lenghtening all strings for no good reason.

Reply:
Yep. This was a pre-existing issue, so it's probably better to merge this PR now, because it fixes a real problem. We can do cleanups later.